src/com/google/ProductionInstrumentationReporter.java

```diff
@@ -0,0 +1,464 @@
+
+        parameterMapping.putIfAbsent(unqiueParam, temp);
+      }
+
+    }
+
+    public String getFileName(String uniqueParam) {
+      return fileNames.get(parameterMapping.get(uniqueParam)[0]);
+    }
+
+    public String getFunctionName(String uniqueParam) {
+      return functionNames.get(parameterMapping.get(uniqueParam)[1]);
+    }
+
+    public InstrumentationType getType(String uniqueParam) {
+      return types.get(parameterMapping.get(uniqueParam)[2]);
+    }
+
+    public int getLineNo(String uniqueParam) {
+      return parameterMapping.get(uniqueParam)[3];
+    }
+
+    public int getColNo(String uniqueParam) {
+      return parameterMapping.get(uniqueParam)[4];
+    }
+
+    /**
+     * This function will return a list of unique parameters which match the criteria specified by
+     * the predicate. As an example, it can return a list of all unique parameters which match a
+     * specific file name.
+     */
+    private List<String> getAllMatchingValues(Predicate<String> comparisonPredicate) {
+      List<String> result = new ArrayList<>();
+      for (String key : parameterMapping.keySet()) {
+        if (comparisonPredicate.test(key)) {
+          result.add(key);
+        }
+      }
+      return result;
+    }
+
+  }
+
+  /**
+   * The instrumentation will send the report in a JSON format where the JSON is a dictionary of the
+   * encoded params to an object of the data collected. Initially this will just be the frequency.
+   */
+  private static class ProfilingData {
+
+    public int frequency;
+
+    public ProfilingData(int frequency) {
+      this.frequency = frequency;
+    }
+
+  }
+
+  /**
+   * This class contains a detailed breakdown of each instrumentation point. For any encoded param,
+   * this class will contain its details. The `executed` property will be a percent of total times
+   * the instrumentation point was called across all provided instrumentation reports. i.e if
+   * this.executed = 100, then every report would have this point present. Similarly, ProfilingData
+   * takes the average of all ProfilingData's for this instrumentation point in all reports.
+   */
+  private static class ProfilingResult {
+
+    public String param;
+    public InstrumentationType type;
+    public int lineNo;
+    public int colNo;
+    public float executed;
+    public ProfilingData data;
+
+    ProfilingResult(InstrumentationMapping instrumentationMapping, String param) {
+      this.param = param;
+      this.type = instrumentationMapping.getType(param);
+      this.lineNo = instrumentationMapping.getLineNo(param);
+      this.colNo = instrumentationMapping.getColNo(param);
+    }
+
+  }
+
+  /**
+   * A class which groups profiling results by source file and contains a mapping of them to each
+   * function which is present in that source file.
+   */
+  private static class ProfilingResultByFile {
+
+    public String fileName;
+    Map<String, List<ProfilingResult>> profilingDataPerFunction;
```

Comment:
Key is function name. Why value is a list instead of single value? Does each value corresponds to a separate report file?